lib/graphql_metrics/analyzer.rb

```diff
@@ -0,0 +1,193 @@
+      end
+
+      def before_query(query)
+        return if query.context[:skip_graphql_metrics_analysis]
+        return unless query.valid?
+
+        query.context.namespace(CONTEXT_NAMESPACE).tap do |ns|
+          ns[RUNTIME_METRICS_ENABLED] = true
+          ns[FIELD_TIMING_CALLBACKS_KEY] = []
+          ns[QUERY_START_TIME_KEY] = current_time
+          ns[QUERY_START_TIME_MONOTONIC_KEY] = current_time_monotonic
+          ns[GRAPHQL_TRACING_FIELD_KEY] = {}
+          ns[GRAPHQL_TRACING_LAZY_FIELD_KEY] = {}
+        end
+      end
+
+      def after_query(query)
+        return if query.context[:skip_graphql_metrics_analysis]
+        return unless query.valid?
+
+        query.context.namespace(CONTEXT_NAMESPACE).tap do |ns|
+          duration = current_time_monotonic - ns[QUERY_START_TIME_MONOTONIC_KEY]
+          runtime_metrics = { start_time: ns[QUERY_START_TIME_KEY], duration: duration }
+
+          analyzer = ns[ANALYZER_INSTANCE_KEY]
+          analyzer.extract_query(runtime_metrics: runtime_metrics, context: query.context)
+          analyzer.run_field_timing_callbacks
+        end
+      end
+
+      def trace(key, data, &block)
+        skip_tracing = data[:query]&.context&.fetch(:skip_graphql_metrics_analysis, false)
+        return block.call if skip_tracing || ![GRAPHQL_TRACING_FIELD_KEY, GRAPHQL_TRACING_LAZY_FIELD_KEY].include?(key)
+        # NOTE: We can't just check `runtime_metrics_enabled?` here, since .trace runs before `before_query`, i.e.
+        # during lexing, parsing, validation etc., and `before_query` doesn't run until `execute_multiplex`.
+
+        if runtime_metrics_enabled?(data[:query]&.context)
+          trace_field(key, data, block, key)
+        else
+          block.call
+        end
+      end
+
+      def trace_field(_key, data, block, context_key)
+        path_excluding_numeric_indicies = data[:path].select { |p| p.is_a?(String) }
+
+        start_time = current_time
+        start_time_monotonic = current_time_monotonic
+
+        result = block.call
+        duration = current_time_monotonic - start_time_monotonic
+
+        data[:query].context.namespace(CONTEXT_NAMESPACE).tap do |ns|
+          ns[context_key][path_excluding_numeric_indicies] ||= []
+          ns[context_key][path_excluding_numeric_indicies] << { start_time: start_time, duration: duration }
+        end
+
+        result
+      end
+
+      def runtime_metrics_enabled?(context)
+        return false unless context
+        !!context.namespace(CONTEXT_NAMESPACE)[RUNTIME_METRICS_ENABLED]
+      end
+    end
+
+    def initialize(query_or_multiplex)
+      super
+
+      query_or_multiplex.context.namespace(CONTEXT_NAMESPACE).tap do |ns|
+        ns[ANALYZER_INSTANCE_KEY] = self
+      end
+
+      @query = query_or_multiplex
+      @static_query_metrics = nil
+      @field_timing_callbacks = []
+    end
+
+    def analyze?
+      query.valid? && query.context[:skip_graphql_metrics_analysis] != true
```

Comment:
@eapache commented about this in another place, but the analyzer is only run if the query is valid anyway. Unless you want this if someone manually runs the analyzer? But that's an odd edge case to account for

Reply:
Yeah I initially decided to support the ad hoc use case, since it seems fairly useful to be able to run extract metrics out of band from production queries. This cached query validity checks, as well the calls to `timings_capture_enabled?` are there to support the analyzer being invoked directly.